### DISCUSSION TITLE
In computing the eigenvalues and eigenvectors of a symmetric rank-2 tensor, the PETSC LAPACK syev routine returned error code 2

### DISCUSSION THREAD

**Discussion by ritamcj** [idaholab/moose#20511] [Q&A General] [answered]
Dear Moose users,
I am running a simulation using tensor mechanics and am seeing this error repeatedly. I tried to reduce the time step but it didn't help. I searched and found one relevant discussion  https://github.com/idaholab/moose/discussions/16288.  My preconditioning and executioner blocks are as follows:
```
[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]

  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.01
    growth_factor = 1.0000001
    cutback_factor = 0.9
    optimal_iterations = 15
    iteration_window = 2
    reject_large_step = true
    reject_large_step_threshold = 0.1
  [../]


  type = Transient
  nl_max_its = 15
  scheme = bdf2
  solve_type = PJFNK
  petsc_options_iname = -pc_type
  petsc_options_value = asm
  l_max_its = 40
  l_tol = 1.0e-3
  nl_rel_tol = 1.0e-5
  start_time = 0.0
  num_steps = 1000
  nl_abs_tol = 1e-5
[]
```

I am running my executable in parallel using mpi as `mpiexec -n 4 ../../../.././drx__ti64-opt -i beta_transformed_Ti64_case1.i`. I will be grateful for any guidance or references regarding this issue.
Warm Regards,
Ritam Chatterjee

**Comment by GiudGiud** [comment, marked as answer]
Hello

Did you try to change a few parameters and see if the error goes away?
Like running it in serial, using lu for preconditioning etc

Guillaume

**Reply by jiangwen84** [reply]
We use LAPACK subroutine at several places in TensorMechanics kernels/materials. It has nothing to do with the Solver option. Very likely, your problem has unrealistic deformation that causes eigendecomposition fail. Could you track where does the error come from?

**Reply by ritamcj** [reply]
Hi Guillaume, yes I am trying `automatic_scaling = true`,`petsc_options_iname = -snes_ksp_ew` according to this webpage https://mooseframework.inl.gov/source/executioners/Steady.html. I will try lu with serial execution and other options and let you know. 

Hi Wen, thank you for the response. Such failure is occurring during elastic deformation. I checked dislocation density, strain tensor components and Cauchy stress but the values seem alright. I will check other tensors such as Fp,Fe and let you know. I am simulating uniaxial compression for Ti6Al4V with beta phase at alpha grain boundaries. Is there anything else I can do to track this error?

**Reply by hugary1995** [reply]
I forgot what the name of that parameter is, but were you using `EigenSolution` or `TaylorExpansion`? Try switching to `TaylorExpansion` and see if that error persists.

**Reply by hugary1995** [reply]
I found it:
```python
# decomposition_method = EigenSolution
decomposition_method = TaylorExpansion
```

**Reply by ritamcj** [reply]
Hi Gary, thank you for your reply. I am using ComputeFiniteStrain, which considers TaylorExpansion as default. However, from your and Wen's suggestion, I got the idea of adding `use_displaced_mesh = true` to all auxkernels and user objects in my input file as I had been facing convergence issues with the progress of plastic deformation. This seems to have helped definitely. Additionally, based on Guillaume's advice, I tried running in serial with `lu` and noticed good improvement in speed. Subsequently, I have run in parallel after making changes to the executioner block as follows:
```
[Executioner]

  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.001
    growth_factor = 1.0000001
    cutback_factor = 0.9
    optimal_iterations = 15
    iteration_window = 2
    reject_large_step = true
    reject_large_step_threshold = 0.1
  [../]


  type = Transient
  nl_max_its = 15
  scheme = bdf2
  solve_type = PJFNK
  petsc_options = -snes_ksp_ew
  petsc_options_iname = '-pc_type -sub_pc_type'
  petsc_options_value = 'asm ilu'
  l_max_its = 40
  l_tol = 1.0e-3
  nl_rel_tol = 1.0e-5
  start_time = 0.0
  num_steps = 20000
  nl_abs_tol = 1e-5
  automatic_scaling = true
  scaling_group_variables = 'disp_x disp_y disp_z'
[]

```
Overall, the simulation is much faster now. However, I've again run into the same PETSC LAPACK error but after significantly more number of time steps. I will explore other options and check.

Thanks and regards,
Ritam

**Reply by hugary1995** [reply]
You could also try the `ElementJacobianDamper` which prevents over distortion of elements. I've had good experience using this damper.

**Reply by ritamcj** [reply]
Hi Gary, I will try `ElementJacobianDamper` and let you know. As a test, I added `decomposition_method = EigenSolution` to the tensor mechanics Master action block in my input file. Now, I haven't seen the error after almost ~15000 time steps with initial dt = 0.005. However, dt has now decreased to 1.5e-5. Is such a decrease expected in multiphase polycrystal plasticity problems?

**Reply by GiudGiud** [reply]
@ritamcj any update on using the `ElementJacobianDamper`
How is the time step decreasing? Is this just the IterationAdaptiveDT decreasing it?

You should play with the parameters of this time stepper. If the system never becomes 'easy' to solve, then it will continuously cut the time step.

**Reply by ritamcj** [reply]
Hi Guillaume,
I was able to figure out the problem. The time step was decreasing due to improper node set generation using `BoundingBoxNodeSetGenerator` as explained in this thread https://github.com/idaholab/moose/discussions/20581. I did try `ElementJacobianDamper` but faced convergence issues, which I can't clearly remember now. I'll check again. I modified my Executioner block and now the simulation is significantly faster (4 times) than earlier when I run using MPI on 8 cores. The executioner block is below:
```
[Executioner]

  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.05
    growth_factor = 1.0000001
    cutback_factor = 0.9
    optimal_iterations = 15
    iteration_window = 2
    reject_large_step = true
    reject_large_step_threshold = 0.1
  [../]


  type = Transient
  nl_max_its = 15
  scheme = bdf2
  solve_type = PJFNK
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter'
  petsc_options_value = '201                hypre    boomeramg      4'
  l_max_its = 40
  l_tol = 1.0e-3
  nl_rel_tol = 1.0e-5
  start_time = 0.0
  num_steps = 20000
  nl_abs_tol = 1e-5
  line_search = none
  snesmf_reuse_base = false
  automatic_scaling = true
  scaling_group_variables = 'disp_x disp_y disp_z'
  [./Predictor]
    type = SimplePredictor
    scale = 1.0
    skip_after_failed_timestep = true
  [../]
[]
```

Warm regards,
Ritam
